### PR TITLE
Added the to_json method to MPD::Song

### PR DESCRIPTION
**Pull request by nounoursheureux**
For a website I'm writing I need to pass a `Song` in JSON format, so I added this method.

### CODE REVIEW

File: lib/ruby-mpd/song.rb [hunk 1]
```diff
@@ -24,6 +24,18 @@ def ==(another)
     self.class == another.class && self.file == another.file
   end
 
+  def to_json(*a)
+    {
```
Comment (archSeer) [inline lib/ruby-mpd/song.rb]: Use symbols as keys, and do a `{ ... }.merge(@data)`. `data` should not be a nested structure.

File: lib/ruby-mpd/song.rb [hunk 2]
```diff
@@ -24,6 +24,18 @@ def ==(another)
...
+      "time" => @time,
+      "file" => @file,
+      "title" => @title,
+      "artist" => @artist,
+      "album" => @album,
+      "albumartist" => @albumartist,
+      "data" => @data
+    }.to_json(*a)
```
Comment (archSeer) [inline lib/ruby-mpd/song.rb]: Also, `to_json` is a rails construct, and as such, this method call will fail on vanilla ruby. Instead, refactor this method into `to_h`, which should return a hash. Then you can call `to_json`/`as_json` on it as needed inside your application (or do any other form of serialization).

Comment (nounoursheureux) [inline lib/ruby-mpd/song.rb]: Are you sure ? `to_json` is mentionned in the official `JSON` [documentation](http://ruby-doc.org/stdlib-2.0.0/libdoc/json/rdoc/JSON.html#module-JSON-label-Generating+JSON)

Comment (whomwah) [inline lib/ruby-mpd/song.rb]: ```to_json``` does just work but would only return ```"\"#<MPD::Song:0x007ff066159ce0>\""```

Would a better implementation be:

```
def to_json
  Hash.new.tap do |hsh|
    instance_variables.each do |var|
      hsh[var] = self.instance_variable_get var
    end
  end.to_json
end
```

As redefining all the instance variables is not ideal.

You could then easily write some test to prove this. You are going to write some tests right?

Comment (archSeer) [inline lib/ruby-mpd/song.rb]: Ah, I stand corrected, `to_json` is indeed valid then.

I would still refactor it by creating a `to_h` implementation and then simply derive the `to_json` implementation from that:

```ruby
def to_h
  {
    time: @time,
    file: @file,
    title: @title,
    artist: @artist,
    album: @album,
    albumartist: @albumartist,
  }.merge(data)
end

def to_json
  to_h.to_json
end
```

Comment (nounoursheureux) [inline lib/ruby-mpd/song.rb]: > to_json does just work but would only return "\"#<MPD::Song:0x007ff066159ce0>\""

This is why I'm redefining it.

> Would a better implementation be:
> 
> def to_json
>   Hash.new.tap do |hsh|
>     instance_variables.each do |var|
>       hsh[var] = self.instance_variable_get var
>     end
>   end.to_json
> end
> As redefining all the instance variables is a bit messy.

I think it is better to explicitely choose which variable should be included in the hash and which shouldn't, so if you happen to have a variable you do not want to include you just have to ignore it.

Comment (whomwah) [inline lib/ruby-mpd/song.rb]: hmm, ok

**Comment by liquid** [comment]
In case you decide to go with adding serialisation to an MPD::Song class, make sure you explicitly define a runtime dependency on the `json` gem. The only reason it works currently is because you are loading all the development dependencies and json extends Hash in in tests.

However, if you ask me personally though, I'd just implement `to_h` and return a hash that the json gem can handle on its own. The [documentation](http://ruby-doc.org/stdlib-2.0.0/libdoc/json/rdoc/JSON.html#module-JSON-label-Generating+JSON) states that...

> an object convertible into a hash by a to_h method

should be sufficient to generate a json structure. That way we don't actually depend on `json` and can get away without having to add another dependency.

/cc @archSeer

**Comment by archSeer** [comment]
@liquid Docs on that are a bit weird, it's just describing `state`, whatever that is (I'm guessing the second parameter, `opts`).

From my testing, to_json doesn't seem to automatically pick up `to_h`.

```
irb(main):001:0> require 'json'
=> true
irb(main):002:0> class A; def to_h; {a: 1}; end end
=> :to_h
irb(main):003:0> A.new.to_h
=> {:a=>1}
irb(main):004:0> A.new.to_json
=> "\"#<A:0x007fbc51a14ce8>\""
```

**Comment by archSeer** [comment]
@nounoursheureux Hmm, alright, let's only define the `to_h` method, we don't really want to include more dependencies in this gem. You can then simply use `to_h.to_json` in your code (and in most cases, an explicit call `to_json` shouldn't be really required, if you're using rails: `render json: song.to_h`). Sorry for all the confusion!

**Comment by nounoursheureux** [comment]
@archSeer Are the tests sufficient ?

**Comment by archSeer** [comment]
@nounoursheureux Everything looks fine! Rebase these commits together please and I'll merge :+1:

**Comment by archSeer** [comment]
Cheers!

**Comment by nounoursheureux** [comment]
:ok_hand:

**Pull request merged**